### PR TITLE
Fixes related to printing tuples, define-fun commands in smt2

### PR DESCRIPTION
**Pull request by ajreynol**
This PR fixes some initial issues with printing datatypes in smt2.

It does not yet address further issues on printing define-fun as a result of `dump=raw-benchmark`, which requires deeper refactoring of the implementation of the API / SmtEngine.

### CODE REVIEW

File: src/printer/smt2/smt2_printer.cpp [hunk 1]
```diff
@@ -951,10 +951,44 @@ void Smt2Printer::toStream(std::ostream& out,
...
+  {
+    Node op = n.getOperator();
+    const DType& dt = DType::datatypeOf(op);
+    if (dt.isTuple())
+    {
+      stillNeedToPrintParams = false;
+      out << "(_ tupSel " << DType::indexOf(op) << ") ";
+    }
+  }
+    break;
```
Comment (mpreiner) [inline src/printer/smt2/smt2_printer.cpp:964]: Is this formatted? Indentation looks wrong.

File: src/printer/smt2/smt2_printer.cpp [hunk 2]
```diff
@@ -951,10 +951,44 @@ void Smt2Printer::toStream(std::ostream& out,
...
+    Node op = n.getOperator();
+    unsigned cindex = DType::indexOf(op);
+    const DType& dt = DType::datatypeOf(op);
+    out << "(_ is ";
+    toStream(out,
+              dt[cindex].getConstructor(),
+              toDepth < 0 ? toDepth : toDepth - 1);
+    out << ") ";
+  }
+    break;
```
Comment (mpreiner) [inline src/printer/smt2/smt2_printer.cpp:977]: Ditto

File: src/printer/smt2/smt2_printer.cpp [hunk 3]
```diff
@@ -951,10 +951,44 @@ void Smt2Printer::toStream(std::ostream& out,
...
+    unsigned index = DType::indexOf(op);
+    const DType& dt = DType::datatypeOf(op);
+    unsigned cindex = DType::cindexOf(op);
+    out << "(_ update ";
+    toStream(out,
+              dt[cindex][index].getSelector(),
+              toDepth < 0 ? toDepth : toDepth - 1);
+    out << ") ";
+  }
+    break;
```
Comment (mpreiner) [inline src/printer/smt2/smt2_printer.cpp:990]: Ditto

File: src/printer/smt2/smt2_printer.cpp [hunk 4]
```diff
@@ -951,10 +951,44 @@ void Smt2Printer::toStream(std::ostream& out,
...
+      stillNeedToPrintParams = false;
+      out << "(_ tupSel " << DType::indexOf(op) << ") ";
+    }
+  }
+    break;
+  case kind::APPLY_TESTER:
+  {
+    stillNeedToPrintParams = false;
+    Node op = n.getOperator();
+    unsigned cindex = DType::indexOf(op);
```
Comment (mpreiner) [inline src/printer/smt2/smt2_printer.cpp:969]:
```suggestion
    uint32_t cindex = DType::indexOf(op);
```

File: src/printer/smt2/smt2_printer.cpp [hunk 5]
```diff
@@ -951,10 +951,44 @@ void Smt2Printer::toStream(std::ostream& out,
...
+    toStream(out,
+              dt[cindex].getConstructor(),
+              toDepth < 0 ? toDepth : toDepth - 1);
+    out << ") ";
+  }
+    break;
   case kind::APPLY_UPDATER:
+  {
+    Node op = n.getOperator();
+    unsigned index = DType::indexOf(op);
```
Comment (mpreiner) [inline src/printer/smt2/smt2_printer.cpp:981]:
```suggestion
    uint32_t index = DType::indexOf(op);
```

File: src/printer/smt2/smt2_printer.cpp [hunk 6]
```diff
@@ -951,10 +951,44 @@ void Smt2Printer::toStream(std::ostream& out,
...
+    const DType& dt = DType::datatypeOf(op);
+    unsigned cindex = DType::cindexOf(op);
```
Comment (mpreiner) [inline src/printer/smt2/smt2_printer.cpp:983]:
```suggestion
    uint32_t cindex = DType::cindexOf(op);
```

File: src/printer/smt2/smt2_printer.cpp (hunk 1 above)
Comment (ajreynol) [reply src/printer/smt2/smt2_printer.cpp:964]: Yep, this is formatted.

**Review by mpreiner — APPROVED** [review]

**Pull request merged**